third_party/move/move-vm/types/src/ty_interner.rs

```diff
@@ -0,0 +1,563 @@
+        results: TypeVecId,
+    },
+}
+
+struct InternMap<T, I> {
+    interned: HashMap<T, I>,
+    data: Vec<T>,
+}
+
+impl<T, I> Default for InternMap<T, I> {
+    fn default() -> Self {
+        Self {
+            interned: HashMap::new(),
+            data: Vec::with_capacity(16),
+        }
+    }
+}
+
+impl<T, I> InternMap<T, I> {
+    fn clear(&mut self) {
+        self.interned.clear();
+        self.data.clear();
+    }
+}
+
+/// Interns single types.
+struct TypeInterner {
+    inner: RwLock<InternMap<TypeRepr, TypeId>>,
+}
+
+impl Default for TypeInterner {
+    fn default() -> Self {
+        Self {
+            inner: RwLock::new(InternMap::default()),
+        }
+    }
+}
+
+impl TypeInterner {
+    fn intern(&self, repr: TypeRepr) -> TypeId {
+        if let Some(id) = self.inner.read().interned.get(&repr) {
+            return *id;
+        }
+
+        let mut inner = self.inner.write();
+        if let Some(id) = inner.interned.get(&repr) {
+            return *id;
+        }
+
+        let id = TypeId(inner.data.len() as u32);
+        inner.data.push(repr);
+        inner.interned.insert(repr, id);
+        id
+    }
+}
+
+/// Interns vector of types (e.g., list of type arguments).
+struct TypeVecInterner {
+    inner: RwLock<InternMap<Arc<[TypeId]>, TypeVecId>>,
+}
+
+impl Default for TypeVecInterner {
+    fn default() -> Self {
+        Self {
+            inner: RwLock::new(InternMap::default()),
+        }
+    }
+}
+
+impl TypeVecInterner {
+    fn intern(&self, tys: &[TypeId]) -> TypeVecId {
+        if let Some(id) = self.inner.read().interned.get(tys) {
+            return *id;
+        }
+
+        let tys_arced: Arc<[TypeId]> = Arc::from(tys);
+        let tys_arced_key = tys_arced.clone();
+
+        let mut inner = self.inner.write();
+        if let Some(id) = inner.interned.get(tys) {
+            return *id;
+        }
+
+        let id = TypeVecId(inner.data.len() as u32);
+        inner.data.push(tys_arced);
+        inner.interned.insert(tys_arced_key, id);
+        id
+    }
+
+    fn intern_vec(&self, tys: Vec<TypeId>) -> TypeVecId {
+        if let Some(id) = self.inner.read().interned.get(tys.as_slice()) {
+            return *id;
+        }
+
+        let tys: Arc<[TypeId]> = tys.into();
+        let tys_key = tys.clone();
+
+        let mut inner = self.inner.write();
+        if let Some(id) = inner.interned.get(&tys) {
+            return *id;
+        }
+
+        let id = TypeVecId(inner.data.len() as u32);
+        inner.data.push(tys);
+        inner.interned.insert(tys_key, id);
+        id
+    }
+}
+
+/// Stores all context for type interning. Users can query interned representations ([TypeId] for
+/// single types or [TypeVecId] for vector of types) based on provided runtime types. Context does
+/// not manage memory nor limit the number of types to intern - this has to be managed externally
+/// by the client (to ensure eviction of interned data is safe).
+pub struct TypeContext {
```

Comment:
Perhaps 'type context' is a bit to generic, `TypeInternContext` is not too long? Elsewhere, we call something like this XXXPool, so `TypePool`. 

Otherwise nice to see this new module.

Reply:
Renamed to `InternedTypePool`?